### ISSUE TITLE
SLR/Moving to error does not handle outgoing messages properly in v6

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus#3032]
taskforce: @MarcinHoppe  @weralabaj @mikeminutillo

Chicken: @tmasternak

# Problem

The goal of this task is to fix two issues identified in Recoverability features in current development version of v6.
* Incorrect logic for moving message to timeouts or error queues. Currently SLR and Faults reuse the receive transaction on the transport level to move messages to timeouts and errors. This transaction gets committed which may (and usually will) result in side-effects of message processing being committed as well e.g. any messages which where produced by handler.
* Current implementation of FLR errors (stored in memory structure which is cleaned as a whole each 5 min) may result in consistently failing message being processed infinitely many times. For example in side-by-side deployment when two instances of endpoint consume message from the same queue. In such case it is possible that instance A will do message processing until it fails and store FLR counter for it in memory. The message will be picked up by instance B which will behave in the same way but before it returns message to queue instance A FLR error data will already be cleared. In fact that will happen also for single instance if rolling back the transaction will take more than 5 min.

# Redesign approach 
The first problem can be solved by the approach taken in v5 i.e. SLR and Faults handling will happen as soon as message reaches SLR or Faults behaviors and will be executed if FLR error counter has reached maximum for SLR to be executed or SLR counter (stored in message header) has reached maximum for Faults logic to be executed.
The second problem will be solved by implementing LRU cache for FLR errors with relatively big constraint on elements count (e.g. 1000). In such case we will prevent any memory leaks from happening and we will minimize the possibility of infinite message processing (for FLR message counter to be cleared 1000 other unique failing messages would have to be observed before reaching maximum FLR retires for the original one). The size of LRU cache should take into consideration the size of the exception object which is stored for each failing message.
Recoverability implementation should not be reused for satellites as their requirements are usually different. In addition for satellites some recoverability features should not be enabled and design should prevent from accidentally turning them on e.g. SLR for .Timeouts. 

# Impact
Taking into consideration first issues described in the *Problem* section the impact is large.

# Plan of Attack
 - [x] Changes in `Core` [PR](https://github.com/Particular/NServiceBus/pull/3614)
   - [x] Change SLR and Faults behavior to match v5
   - [x] Unit tests for transport transaction modes other than `None`: behaviors have to be invoked twice (@tmasternak)
   - [x] Merge FLR failure storage with SLR/MoveToErrorQueue failure storage (@weralabaj)
     - [x]  Separate TimeoutManager storage (@weralabaj)
     - [x] Make LRU cache size configurable (@weralabaj)
   - [x] Make sure exception is properly captured before adding to message headers (@MarcinHoppe)
   - [x] Make FLRetries header available for custom SLR policies (@MarcinHoppe)
   - [x] Turn off SLR for transaction mode `None` (@tmasternak)
   - [x] Internal TF Review
 - [x] Replace acceptance tests that cover this bug to component level tests [PR](https://github.com/Particular/NServiceBus/pull/3664)
 - [x] Maintainer group Review
 - [x] Handle large error headers
  - [x] Create `Consider for V6 Launch` issue in PlatDev and link to here and this [bug](https://github.com/Particular/NServiceBus/issues/3023). [Issue](https://github.com/Particular/PlatformDevelopment/issues/699) has been created
 - [x] Documentation (@tmasternak)
  - [x] Review v5 doco wrt FLR/SLR when transactions are turned off. [PR](https://github.com/Particular/docs.particular.net/pull/1365)
- [x] Update doco - not throwing exception (@weralabaj) https://github.com/Particular/docs.particular.net/commit/1b2c815746f27691315803fcfec808cdda8d8d89
 - [x] Investigate failing RabbitMQ ATTs

# Downstream repos:  
 - [x] TimeoutManager [PR](https://github.com/Particular/NServiceBus/pull/3337)
 - [ ] ~~Gateway [issue](https://github.com/Particular/NServiceBus.Gateway/issues/41)(Out of scope for this work)~~

**Comment by tmasternak** [comment]
@danielmarbach @andreasohlund @SzymonPobiega do you think that the new approach looks good enough?

**Comment by weralabaj** [comment]
@tmasternak Should we mention what we discussed regarding satellites with Andreas?

The implementations for satellites need to be adjusted. In case of timeouts we don't use SLR and we don't want users to mess up with our retry logic (e.g. by disabling FLR or changing number of retries). In case of Gateway it makes sense to have a separate set of settings for retry logic, which might be different from the rest of the solution.
As discussed yesterday, seems like copy-pasting is a sensible option. We (me, Tomasz and Andreas) agreed that it makes more sense than the current implementation.

**Comment by tmasternak** [comment]
@weralabaj done

**Comment by tmasternak** [comment]
Updated the description with exception objec size considerations.

**Comment by tmasternak** [comment]
@andreasohlund I think that we need to agree on the cache size and the memory constraints for it we should be should be pretty much ready with the scope.

**Comment by andreasohlund** [comment]
@tmasternak I can't see how that can affect the desing so let's proceed. We can tune the cache since after we have tested the new solution?

**Comment by tmasternak** [comment]
sounds good

**Comment by andreasohlund** [comment]
On a call with @timbussmann and @indualagarsamy we talked about merging the behaviors and also merging the retry policies. The new "retry policy" would look something like this

```
RetryStrategy MyPolicy(Ex, message,RetryState state)
{
    //state.PerformedImmediateRetriesCount
    //state.LastDelayUsed
    //state.PerformedDelayedRetriesCount
       
if(state.PerformedImmediateRetriesCount < 3) 
    return Retry.Immediate;

if(X)
    return Retry.Delayed(state.LastDelayUsed + Timespan.FromSeconds(10)*state.PerformedDelayedRetriesCount);


    return Retry.IGiveUp;
}
```

Note how we maintain and pass the state to the method since keeping that is non trivial and we keep that state for our own usage anyway.

cc @MarcinHoppe

**Comment by mikeminutillo** [comment]
Please update the Plan of Attack on this issue and link to relevant sub-issues in other repos.

@tmasternak @weralabaj @MarcinHoppe

**Comment by MarcinHoppe** [comment]
@mikeminutillo Done. :point_up:

**Comment by DavidBoike** [comment]
This issue has been identified as a blocker for a limited beta (including only MSMQ/NH). Could someone from the TF please give a status that includes which items in the PoA must be completed before we could send the first email to a potential beta customer?

**Comment by tmasternak** [comment]
@DavidBoike All code items from PoA are done as a single PR for this one. So we need the PR to be merged. Internal TF review will be finished tomorrow and after that it will go through maintainers review. I think there is a good chance we will finish the whole issue at the beginning of the next week. 

That being said if we don't want to deliver beta package sooner or run some tests that depend on this issue my opinion is we can sent the email now. @timbussmann @andreasohlund can you validate if I'm not missing anything here?

**Comment by andreasohlund** [comment]
>That being said if we don't want to deliver beta package sooner or run some tests that depend on this issue my opinion is we can sent the email now. @timbussmann @andreasohlund can you validate if I'm not missing anything here?

:+1:

**Comment by tmasternak** [comment]
I'm changing my role to `Chicken`. Do we think we can include someone to fill my place? Maybe @bording since he is different time-zone?

**Comment by mikeminutillo** [comment]
@tmasternak I will join if there is anything left to do. This puts me on 3 PlatDev task forces but allows us to close off https://github.com/Particular/PlatformDevelopment/issues/701 which is one of them.

**Comment by tmasternak** [comment]
Status update: As a result of maintainers review we are discussing the potential to change integration testing into component testing. The decision will be made today and reflected in the work on the PR.

**Comment by MarcinHoppe** [comment]
PoA updated after second meeting today.

**Comment by MarcinHoppe** [comment]
Testing PR still in flight

**Comment by tmasternak** [comment]
@MarcinHoppe should be moved  in progress as well

**Comment by MarcinHoppe** [comment]
@tmasternak thanks. Hard to do on a phone :).

**Comment by tmasternak** [comment]
We have merged the main PR. Currently the test PR is under review.

**Comment by MarcinHoppe** [comment]
I updated the PoA to reflect this.

**Comment by mikeminutillo** [comment]
I'm not sure which Rabbit ATT errors are being referred to in the PoA. Is `develop` up to date for Rabbit? If so, [those tests are passing fine now](https://builds.particular.net/viewLog.html?buildId=157867&tab=buildResultsDiv&buildTypeId=NServiceBusRabbitMQ_2RunAcceptanceTests). What else is there to do here?

**Comment by tmasternak** [comment]
@mikeminutillo those were random failures but @bording confirmed we are good.

**Comment by MarcinHoppe** [comment]
Finally closed. :dancer:
